### ISSUE TITLE
Refactor packet pipeline

### ISSUE THREAD

**Issue by GrizzlT** [GrizzlT/FalconMC#14]
### Solutions
- **Buffer size**: We use 4 buffers per connection, the default sizes of these buffers are 4kb. We currently shrink as quickly as possible after we exceed this size. Per connection there are two fixed buffers, one for writing data to and one for reading data to. The other two buffers are allowed to grow in size. Between the "fixed" and "dynamic" buffer, (de)compression may take place.

- **New/Better traits**: The mentioned six traits have been implemented for a number of base cases. **The size function should return an exact size in bytes!!** In cases where the packet contents do not depend on the size of the data, it does not have to be exact, but this tends to introduce bugs that are hard to track. If you are in the extremely rare case that you don't give an exact result, it is more efficient to give an upper limit. (This really should be avoided at all times!)

- **Packet Caching**: We'll hold off on this feature for a later time. [`moka`](https://crates.io/crates/moka) has been selected as caching library, we will reintroduce this in a new pull request.

- **Compression and encryption**: Compression is fully supported although not used. Encryption requires authentication with mojang so we will also reintroduce this feature in a new pull request.

- **Minor detail(s)**: `falcon_send!` and `falcon_receive!` are now back to being attribute macros! yay!

<br><br>

### Problem
The current network pipeline that's implemented in Falcon isn't entirely optimal.

The (only?) example of being quite optimal for performance are the current attribute macros `falcon_send!` and `falcon_receive!`.
These expose a single function that then dispatches to the correct packet's writing or reading process according to protocol state and packet id without any more allocations and return as early as possible.

However, for the rest of the pipeline, at least the following should ideally be improved upon:

- **Buffer size**: We should strictly guard the buffer sizes used for every connection, I think these should be capped at 4kb or 8kb.
    When large packets need to be processed, these buffers may grow but should always be shrunk after. How to do this is still an open question that needs answering. ***`?`*** Ideally we should also implement a metrics system to find out the average packet size sent by minecraft clients.

- **New/Better traits**: The current `PacketEncode` and `PacketDecode` traits aren't that well-crafted. These should be refactored into a much more modular system. I suggest using 6 traits (in 2 sets of 3) to do this (inspired by serde):
    - `PacketRead`, `PacketWrite` and `PacketSize`: Packet traits implemented by self-describing types (i.e. types that don't need extra data to be read or written, such as integers, uuid, boolean, ...).
    - `PacketReadSeed`, `PacketWriteSeed` and `PacketSizeSeed`: Packet traits implemented by types that need data not
        provided by the data stream or the type itself (e.g. strings have a maximum length that must be checked
        when reading or writing, `Vec`s usually need to be fed with another field in order to know how many elements they consist of).

    `PacketSize` and `PacketSizeSeed` may be required to either be exact or give an accurate estimate. I would rather choose the requirement of exactness but that is up to context and efficiency concerns. ***`?`***\
    Alongside with this, the derive macros for these traits should be improved by a lot. Questions I have about these
    is how to deal with enums: How do enums appear in the minecraft protocol? ***`?`*** How should we deal with these? ***`?`***\
    And then the details of the macro itself, those belongs in a different issue however.

- **Packet caching**: How do we want this? I think we should only cache expensive packets
    that don't update constantly (such as world chunks). How this would work in the new pipeline I don't know yet. ***`?`***

- **Compression and encryption**: At some point we're also gonna need to deal with this, better to think
    about this early on. I did some research on this and it turns out streaming these two processes is very feasible.
    Question remains: How do they fit in our pipeline? Especially, how would caching would work with compression enabled? ***`?`***

- **Minor detail(s)**: While the current `falcon_send!` and `falcon_receive!` macros are quite optimal for performance,
    they're not the most maintainer-friendly to work with. I made the mistake earlier of going from a macro that
    works on a module declaration to a function-like macro which doens't provide the best dev-experience for most IDEs.

### To summarize
I think #21 is ready to be merged. Documentation still lacks for now.
~~Lots of questions that need to be answered first, I would love to hear everyone's opinion on this.~~

**Comment by GrizzlT** [comment]
I have been thinking a while about these points.

Here are my thoughts:

- **Buffer size**: It was easier for me to think about receiving than sending. I drew up some diagrams over at [the wiki](https://wiki.falconmc.org/dev/falconmc/developer/networking.html) where it's explained in more detail.
    My main question about this right now is whether we *actively* shrink as soon as we can, or whether we *passively* shrink only when the new data doesn't immediately require a reallocation.

- **New Traits**: I have a prototype ready for the mentioned six traits and in the workings of writing up a derive macro to go along with the main 3 ones. These look promising.

- **Compression & Encryption**: I think these are fairly feasible to implement, I'll draw up a diagram soon on the wiki and update this comment by then.

- **Minor details**: We should totally open a pr to fix that dev-unfriendly feature!!

- **Packet caching**: I don't have an answer to this yet, do we want to cache packet compressed (very few ideas) or uncompressed (might be a little easier, but is it as good?).
    If we want to compress them before we cache packets, we should come up with a central caching system that is made sure to be cleared once in a while.
    If we want to not compress them before we cache packets, it might be easier to implement quickly, but performance may take a hit this way. (Plus first option is just more ideal if it works well imo)
